Promotion has many codes

philbirt opened the pull request:
This is still a work in progress, but wanted to get some eyes on it.

First commit:
* Initial modeling and migrations. We will need to get this out first, before all of the other bigger logical pieces. Between this commit and the other, we will need to run a rake task to migrate all the associations over so that when we do launch it, everything will be hooked up correctly. There is a migration in here to provide a clear upgrade path for other uses, but for our case a custom rake task seems more viable.

Second commit:
* Usage limits and other logic will be pushed down to the child codes for every promotion. Searching in admin will use the promotion codes table. Usages of promotions will start populating the new fields on adjustments and order_promotions tables.

Still WIP:
* Admin panel for creating promotions
  * With singular codes
  * With many codes
  * With no codes (site wide)
* More tests around usage limit on the per-promotion level (site wide)

jordan-brough commented on core/db/migrate/20131211112807_create_spree_orders_promotions.rb:
```diff
@@ -1,8 +1,8 @@
-      t.references :order
+	def change
+		create_table :spree_orders_promotions, :id => false do |t|
+			t.references :order
```
accidental changes?

philbirt replied:
Yes, reverted. Thanks!

athal7 commented on backend/app/views/spree/admin/promotions/_form.html.erb:
```diff
@@ -41,7 +56,7 @@
       <%= f.label :usage_limit %><br />
       <%= f.number_field :usage_limit, :min => 0, :class => 'fullwidth' %><br>
```
does this then update all of the promotion_codes' usage limits ?

philbirt replied:
Yeah, I think that would be the idea. In the case of site-wide promos (no associated codes), it would just update it's own usage_limit I think.

athal7 commented:
Wow there's a whole lot to this feature. Impressive job peeling it apart.

jordan-brough commented:
@gmacdougall @cbrunsdon @jhawthorn I haven't taken a close look at it myself yet but it would be great to get FRT thoughts on this.

gmacdougall commented on backend/app/views/spree/admin/promotions/_form.html.erb:
```diff
@@ -23,6 +18,26 @@
       <% end %>
     </div>
 
+    <% if @promotion.promotion_codes.present? %>
```
It's probably more appropriate to use:

```
@promotion.promotion_codes.present?
```

gmacdougall commented on backend/app/views/spree/admin/promotions/_form.html.erb:
```diff
@@ -23,6 +18,26 @@
     </div>
 
+    <% if @promotion.promotion_codes.present? %>
+      <%= f.label :promotion_codes %><br />
```
Rather than a line break, I'd prefer this be dealt with using CSS. Very minor, but will make it easier to restyle large chunks of the admin by not relying on formatting in the markup.

gmacdougall commented on core/app/models/spree/order_promotion.rb:
```diff
@@ -0,0 +1,9 @@
+    belongs_to :order, class_name: 'Spree::Order'
+    belongs_to :promotion, class_name: 'Spree::Promotion'
+    belongs_to :promotion_code, class_name: 'Spree::PromotionCode'
+  end
```
This model should contain validations.

gmacdougall commented on core/app/models/spree/promotion.rb:
```diff
@@ -11,36 +11,63 @@ class Promotion < ActiveRecord::Base
+    scope :applied, -> do
+      joins("INNER JOIN #{order_join_table} ON " \
+        "#{order_join_table}.promotion_id = #{table_name}.id").distinct
+    end
```
I'd prefer that this be built using AREL to ensure cross DB compatibility.

gmacdougall commented on core/app/models/spree/promotion.rb:
```diff
@@ -11,36 +11,63 @@ class Promotion < ActiveRecord::Base
+        "#{order_join_table}.promotion_id = #{table_name}.id").distinct
+    end
+
+    scope :coupons, -> do
```
I don't like this scope. It does more than what it would indicate by name.

philbirt replied:
I dont think this is being used anywhere, I pulled it from the surfdome PR. I'll remove it.

gmacdougall commented on core/app/models/spree/promotion.rb:
```diff
@@ -11,36 +11,63 @@ class Promotion < ActiveRecord::Base
+    end
+
+    def promotion_code(coupon_code)
+      codes.where(Spree::PromotionCode.arel_table[:value].matches("%#{coupon_code}")).first
```
Why matches `%#{coupon_code}`?

philbirt replied:
This is for case-insensitivity of the value field, so we can look it up regardless of the casing.

gmacdougall commented on core/app/models/spree/promotion.rb:
```diff
@@ -57,22 +84,18 @@ def activate(payload)
 
-      return action_taken
+      # If an action has been taken, report back to whatever activated this promotion.
+      return results.include?(true)
```
```
results.any?
```

philbirt replied:
Clever, changing

gmacdougall commented on core/app/models/spree/promotion_handler/cart.rb:
```diff
@@ -38,13 +40,19 @@ def promotions
           ).join_sources
 
           Promotion.active.includes(:promotion_rules).
+            joins('LEFT JOIN "spree_promotion_codes" on "spree_promotions"."id" = "spree_promotion_codes"."promotion_id"').
```
I would prefer that this be build with AREL for cross DB compatibility.

gmacdougall commented on core/app/models/spree/promotion_handler/free_shipping.rb:
```diff
@@ -20,11 +20,12 @@ def activate
-            :path => nil
-          })
+          Spree::Promotion.
+            joins('LEFT JOIN "spree_promotion_codes" on "spree_promotions"."id" = "spree_promotion_codes"."promotion_id"').where({
```
Build with AREL for cross DB compatibility.

gmacdougall commented on core/db/migrate/20150113002122_create_spree_promotion_codes.rb:
```diff
@@ -0,0 +1,13 @@
+    create_table :spree_promotion_codes do |t|
+      t.references :promotion, index: true
+      t.string :value
+      t.integer :usage_limit
```
Promotion and Value should be not_null.

gmacdougall commented on core/db/migrate/20150113141424_move_promotion_codes_to_spree_promotion_code.rb:
```diff
@@ -0,0 +1,18 @@
+    end
+  end
+
+  def down
```
This should be completed.

philbirt replied:
I'm not entirely sure how we could make this backwards compatible. How could we determine which promotions had their codes created because of this migration and which promotions/codes were created after the `up` had been run? We could just drop everything in the promotion_codes table I suppose, but that seems pretty heavy handed

gmacdougall commented on core/db/migrate/20150113002123_create_adjustment_promotion_code_association.rb:
```diff
@@ -0,0 +1,6 @@
+class CreateAdjustmentPromotionCodeAssociation < ActiveRecord::Migration
+  def change
```
Indentation Issues

gmacdougall commented on core/db/migrate/20150213160148_add_promotion_code_id_to_orders_promotions.rb:
```diff
@@ -0,0 +1,8 @@
+class AddPromotionCodeIdToOrdersPromotions < ActiveRecord::Migration
+  def change
```
Indentation Issues

gmacdougall commented on core/db/migrate/20150113141424_move_promotion_codes_to_spree_promotion_code.rb:
```diff
@@ -0,0 +1,18 @@
+    	if promo.read_attribute(:code).present?
+      	promo.codes.create!(value: promo.read_attribute(:code), usage_limit: promo.usage_limit)
+
+      	promo.actions.each do |action|
```
Indentation Issues

gmacdougall commented:
I have some quick feedback on stylistic issues.

I feel there are also a few major problems:

- Substantial chunks of code were added without appropriate test coverage (especially in `Spree::Promotion`).

- This is too large of a step to take in one go. I imagine there are issues with it in its current state, and I couldn't begin to get a grasp on everything. I'd suggest making this change in smaller chunks. I would probably start with moving code out of `Spree::Promotion` in to its own table and maintaining all API compatibility (i.e. a Spree::Promotion has_one Spree::PromotionCode). That gets you somewhere closer to where you need to be without breaking everything.

- We don't have a good place to be providing documentation for what has changed, which makes it difficult when we make breaking API changes like this.

philbirt commented:
Thanks for taking a look at this, very good feedback. I have gone through most of the stylistic changes and modified the offending snippets. I also deleted a fair chunk of code introduced in `Spree::Promotion` that was both untested and unused, and re-ordered the methods in there so that the diff looks (a little) less insane.

As far as your second point, I think that would definitely be a good first step. We could just delegate the code on promotion to its association, and that would be pretty much it.

The real complicated issue here is that when you have a single promotion with many codes, the usage limit/eligibility logic changes significantly, the biggest change being that the code needs to track its own logic, and `spree_order_promotions` and `spree_adjustments` need to keep track of the `Spree::PromotionCode`, as well as continuing to keep track of the `Spree::Promotion` and `Spree::PromotionAction` respectively. Unfortunately, there isn't any smaller-step way of achieving this goal that I can think of. Once you move to a one-to-many relationship, you have to get out the microscope on any place that uses eligibility or usage limit logic and any place that needs to populate the promotion_code_id.

So when you really break it down, this change doesn't really introduce much in terms of new logic (just delegating usage limit and code to a new association on `Spree::Promotion`). The biggest scary thing in my eyes would be if the spec coverage didn't catch something...which is definitely possible. That said, getting the core specs back to green made it seem like there was fairly adequate coverage of the desired behavior of the system. Between that and a very large amount of QA, which I guarantee is on deck for us to alleviate our own fears of missing something, I'm not sure how we could break this problem down in a significantly different way. At some point we have to trust the specs, and if not, add some new ones.

gmacdougall commented:
Thanks for giving that some thought @philbirt. I agree that there is that large leap in functionality from when a promotion goes from having one code to many codes. My thinking was that if we can isolate that particular change to be as small as possible, it will be easiest to understand when reviewing potential impacts to code. I know that even making that change as small as possible will still be a lot of lines of code, and quite complex.

At the end of the day, we'll still have to put it through the paces as I'm not super confident in the promo system testing as it is. Anything we can do to improve our unit tests here would also be a big win.

jordan-brough commented:
I like @gmacdougall's idea of putting out the table separation as a first step.  That makes things much more rollback-able if something were to go terribly wrong with the logic updates as well. Are you up for doing that @philbirt?

philbirt commented:
Alright, after talking it through with @jordan-brough and hearing @gmacdougall concerns, I will be breaking this big guy up into smaller chunks. The current plan is as follows:

Pull Request/Deploy 1: (https://github.com/bonobos/spree/pull/248)
- migration to create new tables/fields
- keep 1-code-per-promo requirement (temporarily)
- update the code to store promo codes in both promotions and promotion_codes, but continue to use promotions for reading data

Hopefully this will be pretty safe and you would never have to roll back past that.

Pull Request/Deploy 2: (https://github.com/bonobos/spree/pull/249)
- migration to backfill all old data
- keep 1-code-per-promo requirement (temporarily)
- switch to promotion_codes for reading data

Deploy 3-X: (This pull request)
- start making code updates that change behavior (e.g. multiple codes per promo)

What do you think of this? I'm going to start trying to rebase/squash the later commits of this on top of the other two, but it would be great to get some reactions before I get too far into it.

Also to note, on this current branch all the tests in all the sub-directories are green! Huzzah!

philbirt commented:
Closing in favor of:
https://github.com/bonobos/spree/pull/258
https://github.com/bonobos/spree/pull/260
